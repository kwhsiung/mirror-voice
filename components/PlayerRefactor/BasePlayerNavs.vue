<template>
  <nav class="navs">
    <button
      class="button rewind-button"
      @click="handleClickRewind"
    >
      <img
        class="rewind-button__img"
        src="~/assets/img/skip15secback.svg"
        alt="rewind"
      >
    </button>
    <button
      v-show="showTracksNav"
      class="button backward-button"
      @click="handleClickBackward"
    >
      <img
        class="backward-button__img"
        src="~/assets/img/btn_forward_normal.png"
        alt="backward"
      >
    </button>
    <button
      :class="[
        'button',
        'main-button',
        { 'play': isPlaying },
        { 'pause': !isPlaying }
      ]"
      @click="handleClickMain"
    >
      <img
        v-show="!isPlaying"
        class="main-button__img"
        src="~/assets/img/btn_play.png"
        alt="play"
      >
      <img
        v-show="isPlaying"
        class="main-button__img"
        src="~/assets/img/btn_pause.png"
        alt="pause"
      >
    </button>
    <button
      v-show="showTracksNav"
      class="button forward-button"
      @click="handleClickForward"
    >
      <img
        class="forward-button__img"
        src="~/assets/img/btn_backward_normal.png"
        alt="forward"
      >
    </button>
    <button
      class="button fastforward-button"
      @click="handleClickFastforward"
    >
      <img
        class="fastforward-button__img"
        src="~/assets/img/skip15sec.svg"
        alt="fastforward"
      >
    </button>
  </nav>
</template>

<script>
export default {
  props: {
    isPlaying: {
      type: Boolean,
      default: false
    },
    showTracksNav: {
      type: Boolean,
      default: false
    }
  },
  computed: {
    mainButtonNextEvent() {
      return this.isPlaying ? 'pause' : 'play'
    }
  },
  methods: {
    handleClickRewind() {
      this.$emit('rewind')
    },
    handleClickFastforward() {
      this.$emit('fastforward')
    },
    handleClickBackward() {
      this.$emit('backward')
    },
    handleClickForward() {
      this.$emit('forward')
    },
    handleClickMain() {
      this.$emit(this.mainButtonNextEvent)
    }
  }
}
</script>

<style lang="stylus" scoped>
.navs
  display flex
  align-items center

.rewind-button
  height 23px
  cursor pointer
  margin 0 18px 0 0

.backward-button
  height 14px
  cursor pointer
  margin 0 18px 0 0

.main-button
  cursor pointer
  width 28px
  height 28px

.forward-button
  height 14px
  cursor pointer
  margin 0 0 0 18px

.fastforward-button
  height 23px
  cursor pointer
  margin 0 0 0 18px

.button
  cursor pointer
  padding 0
  background-color transparent
  border none
  outline none
  img
    height 100%

@media (max-width 768px)
  .rewind-button
    display none

  .backward-button
    display none

  .main-button
    width 40px
    height 40px

  .forward-button
    height 15px

  .fastforward-button
    display none
</style>
