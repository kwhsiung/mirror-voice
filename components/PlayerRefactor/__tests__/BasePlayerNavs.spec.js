import { shallowMount } from '@vue/test-utils'
import BasePlayerNavs from '../BasePlayerNavs.vue'

describe('BasePlayerNavs', () => {
  test('emit "rewind" event while rewind button clicked', () => {
    const wrapper = shallowMount(BasePlayerNavs)
    const button = wrapper.find('.rewind-button')
    button.trigger('click')

    expect(wrapper.emitted('rewind')).toHaveLength(1)
  })

  test('emit "fastforward" event while fastforward button clicked', () => {
    const wrapper = shallowMount(BasePlayerNavs)
    const button = wrapper.find('.fastforward-button')
    button.trigger('click')

    expect(wrapper.emitted('fastforward')).toHaveLength(1)
  })

  test('emit "backward" event while backward button clicked', () => {
    const wrapper = shallowMount(BasePlayerNavs)
    const button = wrapper.find('.backward-button')
    button.trigger('click')

    expect(wrapper.emitted('backward')).toHaveLength(1)
  })

  test('emit "forward" event while forward button clicked', () => {
    const wrapper = shallowMount(BasePlayerNavs)
    const button = wrapper.find('.forward-button')
    button.trigger('click')

    expect(wrapper.emitted('forward')).toHaveLength(1)
  })

  test('if "showTracksNav" is false, hide container of forward/backward buttons, otherwise show the buttons', () => {
    let wrapper = shallowMount(BasePlayerNavs, {
      propsData: {
        showTracksNav: false
      }
    })
    let buttonForward = wrapper.find('.forward-button')
    let buttonBackward = wrapper.find('.backward-button')

    expect(buttonForward.element.style.display).toBe('none')
    expect(buttonBackward.element.style.display).toBe('none')

    wrapper = shallowMount(BasePlayerNavs, {
      propsData: {
        showTracksNav: true
      }
    })
    buttonForward = wrapper.find('.forward-button')
    buttonBackward = wrapper.find('.backward-button')

    expect(buttonForward.element.style.display).not.toBe('none')
    expect(buttonBackward.element.style.display).not.toBe('none')
  })

  test('if "isPlaying" props is true, main button will be pause button', () => {
    const wrapper = shallowMount(BasePlayerNavs, {
      propsData: {
        isPlaying: true
      }
    })
    const button = wrapper.find('.main-button')

    expect(button.classes()).toContain('play')

    button.trigger('click')

    expect(wrapper.emitted('update:isPlaying')[0]).toEqual([false])
  })

  test('if "isPlaying" props is false, main button will be play button', () => {
    const wrapper = shallowMount(BasePlayerNavs, {
      propsData: {
        isPlaying: false
      }
    })
    const button = wrapper.find('.main-button')

    expect(button.classes()).toContain('pause')

    button.trigger('click')

    expect(wrapper.emitted('update:isPlaying')[0]).toEqual([true])
  })
})

describe('snapshot tests', () => {
  test('renders component correctly if showTracksNav is false', () => {
    const wrapper = shallowMount(BasePlayerNavs, {
      propsData: {
        showTracksNav: false
      }
    })

    expect(wrapper.element).toMatchSnapshot()
  })
  test('renders component correctly if showTracksNav is true', () => {
    const wrapper = shallowMount(BasePlayerNavs, {
      propsData: {
        showTracksNav: true
      }
    })

    expect(wrapper.element).toMatchSnapshot()
  })
})
