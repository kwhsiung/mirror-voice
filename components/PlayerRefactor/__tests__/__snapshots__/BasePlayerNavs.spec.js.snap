// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot tests renders component correctly if showTracksNav is false 1`] = `
<nav
  class="navs"
>
  <button
    class="button rewind-button"
  >
    <img
      alt="rewind"
      class="rewind-button__img"
      src="~/assets/img/skip15secback.svg"
    />
  </button>
   
  <button
    class="button backward-button"
    style="display: none;"
  >
    <img
      alt="backward"
      class="backward-button__img"
      src="~/assets/img/btn_forward_normal.png"
    />
  </button>
   
  <button
    class="button main-button pause"
  >
    <img
      alt="play"
      class="main-button__img"
      src="~/assets/img/btn_play.png"
    />
     
    <img
      alt="pause"
      class="main-button__img"
      src="~/assets/img/btn_pause.png"
      style="display: none;"
    />
  </button>
   
  <button
    class="button forward-button"
    style="display: none;"
  >
    <img
      alt="forward"
      class="forward-button__img"
      src="~/assets/img/btn_backward_normal.png"
    />
  </button>
   
  <button
    class="button fastforward-button"
  >
    <img
      alt="fastforward"
      class="fastforward-button__img"
      src="~/assets/img/skip15sec.svg"
    />
  </button>
</nav>
`;

exports[`snapshot tests renders component correctly if showTracksNav is true 1`] = `
<nav
  class="navs"
>
  <button
    class="button rewind-button"
  >
    <img
      alt="rewind"
      class="rewind-button__img"
      src="~/assets/img/skip15secback.svg"
    />
  </button>
   
  <button
    class="button backward-button"
  >
    <img
      alt="backward"
      class="backward-button__img"
      src="~/assets/img/btn_forward_normal.png"
    />
  </button>
   
  <button
    class="button main-button pause"
  >
    <img
      alt="play"
      class="main-button__img"
      src="~/assets/img/btn_play.png"
    />
     
    <img
      alt="pause"
      class="main-button__img"
      src="~/assets/img/btn_pause.png"
      style="display: none;"
    />
  </button>
   
  <button
    class="button forward-button"
  >
    <img
      alt="forward"
      class="forward-button__img"
      src="~/assets/img/btn_backward_normal.png"
    />
  </button>
   
  <button
    class="button fastforward-button"
  >
    <img
      alt="fastforward"
      class="fastforward-button__img"
      src="~/assets/img/skip15sec.svg"
    />
  </button>
</nav>
`;
